### PR TITLE
8289030: [macos] app image signature invalid when creating DMG or PKG

### PR DESCRIPTION
**Pull request by sashamatveev**
Fixed 3 issues which made signature invalid:
- We should not remove .jpackage.xml from signed app image when creating DMG or PKG otherwise it invalidates signature.
- .package should be created when app image is generated, so this file can be signed.
- Copying predefine app image for DMG and PKG should not follow symbolic links, otherwise several files from runtime (COPYRIGHT and LICENSE) will be copied instead of symbolic links being created, since it invalidates signature as well.

Added additional test to validate signature when DMG or PKG is generated from predefined app image.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8289030](https://bugs.openjdk.org/browse/JDK-8289030): [macos] app image signature invalid when creating DMG or PKG

### Reviewers
 * [Alexey Semenyuk](https://openjdk.org/census#asemenyuk) (@alexeysemenyukoracle - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk19 pull/89/head:pull/89` \
`$ git checkout pull/89`

Update a local copy of the PR: \
`$ git checkout pull/89` \
`$ git pull https://git.openjdk.org/jdk19 pull/89/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 89`

View PR using the GUI difftool: \
`$ git pr show -t 89`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk19/pull/89.diff">https://git.openjdk.org/jdk19/pull/89.diff</a>

</details>

### CODE REVIEW

**Comment by victordyakov** [comment]
@alexey_semenyuk please review

**Comment by alexeysemenyukoracle** [comment]
I think we can do the following for the signed image: don't add the `.package` file. Instead, write a warning saying that because the app image is signed, support for per-user configuration of the installed app will not be working. (https://bugs.openjdk.org/browse/JDK-8287060 refers to per-use configuration)

**Comment by sashamatveev** [comment]
8289030: [macos] app image signature invalid when creating DMG or PKG [v2]
 - Fixed as suggested by Alexey. If app image already signed we will not add `.package` file.

File: test/jdk/tools/jpackage/macosx/SigningPackageTwoStepTest.java [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+    private static void verifyDMG(JPackageCommand cmd) {
+        // DMG always unsigned, so we will check it
+        Path outputBundle = cmd.outputBundle();
+        SigningBase.verifyCodesign(outputBundle, false);
+    }
+
+    private static void verifyAppImageInDMG(JPackageCommand cmd) {
+        MacHelper.withExplodedDmg(cmd, dmgImage -> {
+            Path launcherPath = dmgImage.resolve(Path.of("Contents", "MacOS", cmd.name()));
```
Comment (alexeysemenyukoracle) [inline test/jdk/tools/jpackage/macosx/SigningPackageTwoStepTest.java:87]: I'd replace it with `ApplicationLayout.macAppImage().resolveAt(dmgImage).launchersDirectory(cmd.name())`

**Review by alexeysemenyukoracle — CHANGES_REQUESTED** [review]

File: src/jdk.jpackage/macosx/classes/jdk/jpackage/internal/MacBaseInstallerBundler.java [hunk 2]
```diff
@@ -158,17 +168,19 @@ protected Path prepareAppBundle(Map<String, ? super Object> params)
...
         } else {
             appDir = appImageBundler.execute(params, appImageRoot);
         }
 
-        if (!StandardBundlerParam.isRuntimeInstaller(params)) {
-            new PackageFile(APP_NAME.fetchFrom(params)).save(
-                    ApplicationLayout.macAppImage().resolveAt(appDir));
-           Files.deleteIfExists(AppImageFile.getPathInAppImage(appDir));
-        }
-
```
[the remark this answers is not in the document]
Comment (alexeysemenyukoracle) [reply src/jdk.jpackage/macosx/classes/jdk/jpackage/internal/MacBaseInstallerBundler.java:171]: Oh, right.
Anyways let's keep `.package`-related stuff away from AbstractAppImageBuilder.java, and AppImageBundler.java.

I'd move `.package`-related logic from AbstractAppImageBuilder to MacAppImageBuilder and change
```
public MacAppBundler() {
        setAppImageSupplier(MacAppImageBuilder::new);
        setParamsValidator(MacAppBundler::doValidate);
    }
```
to something like this
```
public MacAppBundler() {
       public MacAppBundler() {
        setAppImageSupplier(imageOutDir -> {
            return new MacAppImageBuilder(imageOutDir, isDependentTask());
        });
        setParamsValidator(MacAppBundler::doValidate);
    }
```

Need to add `AppImageBundler.sDependentTask()` method and change signature if MacAppImageBuilder ctor from `MacAppImageBuilder(Path imageOutDir)` to `MacAppImageBuilder(Path imageOutDir, boolean withPackageFile)`

File: test/jdk/tools/jpackage/macosx/SigningPackageTwoStepTest.java (hunk 1 above)
Comment (sashamatveev) [reply test/jdk/tools/jpackage/macosx/SigningPackageTwoStepTest.java]: Fixed.

File: src/jdk.jpackage/macosx/classes/jdk/jpackage/internal/MacBaseInstallerBundler.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (sashamatveev) [reply src/jdk.jpackage/macosx/classes/jdk/jpackage/internal/MacBaseInstallerBundler.java:171]: Fixed.

**Review by alexeysemenyukoracle — APPROVED** [review]

**Pull request closed without merging**